### PR TITLE
Room assignment

### LINKED ISSUE #2: Room assignment module
Create a page where we can store and admins/secretariat/president can edit which room the students and guests live. Visible to collegists.

A room can have 2 or 3 beds.

### PR DESCRIPTION
**Pull request by sszajbely**

### CODE REVIEW

File: app/Http/Controllers/Dormitory/RoomController.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    {
+        $this->authorize('updateAny', Room::class);
+
+        $validator = Validator::make($request->all(), [
+            'type' => 'required|string|in:add,remove'
+        ]);
+        $validator->validate();
+        $new_capacity=$room->capacity+($request->type=='add' ? 1 : -1);
+        if ($new_capacity<$room->residentNumber()) {
+            return back()->with('error', 'hiba');
```
Comment (kdmnk) [inline app/Http/Controllers/Dormitory/RoomController.php]: Make proper error messages

File: app/Http/Controllers/Dormitory/RoomController.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        }
+        if ($new_capacity>4 || $new_capacity<0) {
+            return back()->with('error', 'masikhiba');
+        }
+        if ($request->type=='add') {
+            $room->increment('capacity');
+            return back();
```
Comment (kdmnk) [inline app/Http/Controllers/Dormitory/RoomController.php]: delete return from here

File: app/Http/Controllers/Dormitory/RoomController.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        } else {
+            $room->decrement('capacity');
+        }
+
+        return back();
+    }
+
+    public function updateResidents(Request $request)
+    {
+        // return $request;
```
Comment (kdmnk) [inline app/Http/Controllers/Dormitory/RoomController.php]: Validate the request, especially if the users exist. There are rules for arrays.

File: app/Models/PersonalInformation.php [hunk 4]
```diff
@@ -24,6 +24,7 @@ class PersonalInformation extends Model
         'street_and_number',
         'tenant_until',
         'profile_picture_id',
+        'room'
```
Comment (kdmnk) [inline app/Models/PersonalInformation.php]: this is not in the personalinfo, right?

File: app/Policies/RoomPolicy.php [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    /**
+     * Determine whether the user can create models.
+     *
+     * @param  \App\Models\User  $user
+     * @return \Illuminate\Auth\Access\Response|bool
+     */
+    public function create(User $user)
+    {
+        //
+    }
```
Comment (kdmnk) [inline app/Policies/RoomPolicy.php]: Delete the ones you do not or plan to use.

File: database/factories/PersonalInformationFactory.php [hunk 6]
```diff
@@ -3,6 +3,7 @@
 namespace Database\Factories;
 
 use App\Models\PersonalInformation;
+use App\Models\Room;
```
Comment (kdmnk) [inline database/factories/PersonalInformationFactory.php]: delete

File: resources/lang/hu/rooms.php [hunk 7]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+return [
+    'rooms' => 'Szobabeosztás',
+    'search' => 'Keresés',
+    'search_name' => 'Keress névre vagy szobaszámra',
+    'resident1' => '1. lakó',
```
Comment (kdmnk) [inline resources/lang/hu/rooms.php:6]: you can do this with an indexed loop, add only 'resident' here

File: resources/views/dormitory/rooms/app.blade.php [hunk 8]
```diff
@@ -0,0 +1,861 @@
...
+                                break;
+                            case 2:
+                                $color="#f0fc0a";
+                                break;
+                            case 3:
+                                $color="#fc4f05";
+                                break;
+                            default:
+                                $color="#ffffff";
+                                break;
```
Comment (kdmnk) [inline resources/views/dormitory/rooms/app.blade.php]: Add the colors to the Room model, something like  in semesterstatus or workshops

File: resources/views/dormitory/rooms/app.blade.php [hunk 9]
```diff
@@ -0,0 +1,861 @@
...
+                        }    
+                        array_push($colorArray, [ "id" => $room->name, "color" => $color]);
+                    }
+                    $roomColors=collect($colorArray);
+                @endphp
+                {{-- {{$users}} --}}
+                {{-- {{$roomColors->where('id',223)->pluck('color')->join(',')}} --}}
```
Comment (kdmnk) [inline resources/views/dormitory/rooms/app.blade.php]: delete comments

File: resources/views/dormitory/rooms/app.blade.php [hunk 10]
```diff
@@ -0,0 +1,861 @@
...
+                <svg
```
Comment (kdmnk) [inline resources/views/dormitory/rooms/app.blade.php]: Could you add a comment what is the process here?

File: resources/views/dormitory/rooms/app.blade.php [hunk 11]
```diff
@@ -0,0 +1,861 @@
...
+                    <g
+                        inkscape:groupmode="layer"
+                        inkscape:label="Layer 2"
+                        style="display:inline"
+                        id="g117">
+                        <path
+                        style="fill: {{$roomColors->where('id',235)->pluck('color')->join(',')}} ;stroke:#000000;stroke-width:0.264583px;stroke-linecap:butt;stroke-linejoin:miter;stroke-opacity:1"
+                        d="m 147.05702,69.05702 0.93545,52.75901 89.80256,-0.18709 -0.37418,-52.94609 z"
+                        id="room235" 
+                        data-info="{{ $users->where('room', 235)->pluck('name')->join(';') }}"/>
```
Comment (kdmnk) [inline resources/views/dormitory/rooms/app.blade.php]: Maybe you can outsource these into a template, or at least to a new file

File: resources/views/dormitory/rooms/app.blade.php [hunk 12]
```diff
@@ -0,0 +1,861 @@
...
+                    id="text11608-27-1"><tspan
+                        class="fix-stroke"
+                        sodipodi:role="line"
+                        id="tspan11606-5-0"
+                        style="font-style:normal;font-variant:normal;font-weight:normal;font-stretch:normal;font-family:Roboto;-inkscape-font-specification:Roboto;stroke:#ffffff;stroke-width:0.882398"
+                        x="661.51129"
+                        y="434.96973">217</tspan></text>
+                    <text
+                    xml:space="preserve"
+                    style="font-style:normal;font-variant:normal;font-weight:normal;font-stretch:normal;font-size:14.1184px;line-height:1.25;font-family:Roboto;-inkscape-font-specification:Roboto;display:inline;stroke-width:0.882398"
```
Comment (kdmnk) [inline resources/views/dormitory/rooms/app.blade.php]: I'm not familiar with SVGs, but probably the style could be outsourced as classes.

File: resources/views/dormitory/rooms/app.blade.php [hunk 13]
```diff
@@ -0,0 +1,861 @@
...
+                    </g>
+                    </svg>
+
+                <div class="row">
+                    <form id="name", onsubmit="return false;" style="vertical-align: top">
+                        <x-input.text s=9  id="nameInput" text="rooms.search_name" type="text" />
+                        <x-input.button s=3 class="right coli blue" text="rooms.search" />
+                    </form>
+                </div>
+                {{-- TODO can --}}
```
Comment (kdmnk) [inline resources/views/dormitory/rooms/app.blade.php]: delete comment

File: app/Http/Controllers/Dormitory/RoomAssignmentController.php [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+use Illuminate\Support\Facades\Auth;
+use Illuminate\Support\Facades\Mail;
+
+class RoomAssignmentController extends Controller
+{
+    public function index()
+    {
+        return view('dormitory.rooms.app');
+    }
+}
```
Comment (kdmnk) [inline app/Http/Controllers/Dormitory/RoomAssignmentController.php]: Please merge the two controllers together.

File: routes/web.php [hunk 15]
```diff
@@ -29,8 +30,11 @@
 use App\Http\Controllers\StudentsCouncil\EconomicController;
 use App\Http\Controllers\StudentsCouncil\EpistolaController;
 use App\Http\Controllers\StudentsCouncil\MrAndMissController;
+use App\Http\Controllers\Dormitory\RoomController;
+use App\Models\Room;
 use Illuminate\Support\Facades\Auth;
 use Illuminate\Support\Facades\Route;
+use App\Models\User;
```
Comment (kdmnk) [inline routes/web.php:36]: Delete unused imports

**Review by kdmnk — CHANGES_REQUESTED** [review]
I have added some ideas on how can you improve code quality.

File: resources/lang/hu/rooms.php (hunk 7 above)
Comment (sszajbely) [reply resources/lang/hu/rooms.php:6]: English won't work that way.

File: app/Models/Room.php [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+
+    public function residentNumber()
+    {
+        return $this->users()->count();
+    }
+
+    public function color()
+    {
+        $color="#ffffff";
+        switch ($this->capacity-$this->residentNumber()) {
```
Comment (kdmnk) [inline app/Models/Room.php]: ??

File: app/Models/User.php [hunk 17]
```diff
@@ -748,4 +748,8 @@ public function votedFor($category): array
 
         return ['voted' => false];
     }
+    public function room()
+    {
+        return $this->belongsTo(\App\Molels\Room::class, 'room', 'name');
```
Comment (kdmnk) [inline app/Models/User.php]: typo
After fixing, check in tinker if this works as expected

**Review by kdmnk — CHANGES_REQUESTED** [review]

File: app/Models/Room.php (hunk 16 above)
Comment (sszajbely) [reply app/Models/Room.php]: It assigns a color by how much capacity is left (maxCapacity-residentsAlreadyThere)

File: resources/views/dormitory/rooms/app.blade.php [hunk 18]
```diff
@@ -0,0 +1,346 @@
...
+                    }
+                    $roomColors=collect($colorArray);
+                @endphp
+                <svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
+                        viewBox="0 0 1047.8 670" style="enable-background:new 0 0 1047.8 750;" xml:space="preserve">
+                    <style type="text/css">
+                        .st0{fill:none;}
+                        .st1{fill:#FFFFFF;stroke:#000000;stroke-width:0.25;}
+                        .st2{fill:#808080;stroke:#000000;stroke-width:0.25;}
+                        .st4{font-family:'RobotoSlab-Regular'; font-size:16px; fill:#000000;stroke:#FFFFFF;stroke-width:0.0; }
```
Comment (kdmnk) [inline resources/views/dormitory/rooms/app.blade.php]: Add pointer-events: none; to st4

File: resources/views/dormitory/rooms/app.blade.php [hunk 19]
```diff
@@ -0,0 +1,346 @@
...
+@section('title')
+<i class="material-icons left">bed</i>@lang('rooms.rooms')
+@endsection
+
+<style>
+
+.fix-stroke {
+   paint-order: stroke fill;
+}
+
```
Comment (kdmnk) [inline resources/views/dormitory/rooms/app.blade.php]: Just add comments on what the functions are doing, how the SVG can be reproduced, the overall concept, etc.

File: app/Models/User.php (hunk 17 above)
Comment (kdmnk) [reply app/Models/User.php]: wtf this is working...

**Review by kdmnk — CHANGES_REQUESTED** [review]
You forgot to add this to the menu. Otherwise, I just left some comments.

File: app/Models/Room.php (hunk 16 above)
Comment (kdmnk) [reply app/Models/Room.php]: Oh, then add spaces please :D

Comment (kdmnk) [reply app/Models/Room.php]: And a comment

**Review by kdmnk — APPROVED** [review]
Merge once you fixed the above.

File: resources/views/dormitory/rooms/app.blade.php (hunk 18 above)
Comment (sszajbely) [reply resources/views/dormitory/rooms/app.blade.php]: I tried so hard to google this one, but didn't find anything... Thanks :D

**Pull request merged**